flocker/node/agents/functional/test_gce.py

```diff
@@ -322,13 +388,91 @@ def test_attach_elsewhere_attached_volume(self):
+        """
+        Verify that if every call to the :class:`GCEAtomicOperation` s is
+        duplicated that we handle the errors correctly.
+
+        This should force some specific scheduling situations that resemble
+        race conditions with another agent trying to converge to the same
+        state, or a condition where the dataset agent as rebooted after a crash
+        that happened in the middle of an :class:`IBlockDeviceAPI` call.
+
+        In these situations we should verify that the second call to many of
+        the underlying atomic methods would result in the correct underlying
+        :class:`VolumeException`.
+        """
+        actual_api = gceblockdeviceapi_for_test(self)
+        volume_manager = actual_api._volume_manager
+        api = actual_api.set(
+            '_volume_manager',
+            repeat_call_proxy_for(IGCEVolumeManager, volume_manager)
+        )
+
+        dataset_id = uuid4()
+
+        # There is no :class:`VolumeException` for creating an already created
+        # volume. Thus, GCE just raises its own custom exception in that case.
+        self.assertThat(
+            lambda: api.create_volume(
+                dataset_id=dataset_id,
+                size=get_minimum_allocatable_size()
+            ),
+            Raises(MatchesException(GCEVolumeException))
+        )
+
+        volumes = api.list_volumes()
+
+        self.assertThat(
+            volumes,
+            AnyMatch(MatchesStructure(dataset_id=Equals(dataset_id)))
+        )
+        volume = next(v for v in volumes if v.dataset_id == dataset_id)
+
+        compute_instance_id = api.compute_instance_id()
+
+        self.assertThat(
+            lambda: api.attach_volume(
+                blockdevice_id=volume.blockdevice_id,
+                attach_to=compute_instance_id,
+            ),
+            Raises(MatchesException(AlreadyAttachedVolume))
+        )
+
+        self.assertThat(
+            api.get_device_path(volume.blockdevice_id).path,
+            Contains('/dev/sd')
+        )
+
+        api.detach_volume(
+            blockdevice_id=volume.blockdevice_id,
+        )
+        self.assertThat(
+            lambda: api.detach_volume(
+                blockdevice_id=volume.blockdevice_id,
+            ),
+            Raises(MatchesException(UnattachedVolume))
+        )
+
+        self.assertThat(
+            lambda: api.destroy_volume(
+                blockdevice_id=volume.blockdevice_id,
+            ),
+            Raises(MatchesException(UnknownVolume))
+        )
+
+        self.assertThat(
+            api.list_volumes(),
+            AllMatch(Not(MatchesStructure(dataset_id=Equals(dataset_id))))
+        )
+
     def test_list_volumes_walks_pages(self):
         """
         Ensure that we can walk multiple pages returned from listing GCE
         volumes.
         """
         api = gceblockdeviceapi_for_test(self)
-        self.patch(api, '_page_size', 1)
+        api = api.set('_page_size', 1)
```

Comment:
so since the API is now a PClass you can't change it, but you can create a new instance that is identical _except_ that it has the `_page_size` attribute set to 1 instead of whatever it previously was set to. `.set()` is pyrsistent's syntactic sugar on that operation. You don't have to worry about setting it back to something else at the end of the test because, well, you have a new instance of the object that only you have a reference to. The old one is still available for whomever else was using it (although in this case the old one has no more references after line 475 (I think pyrsistent has some magic that makes the operation efficient in this case)).